### PR TITLE
[Ready To Merge!]Blob Mode Major Overhaul

### PR DESCRIPTION
**Pull request by Fox-McCloud**
@Iamgoofball expressed interest in porting over /TG/'s blob over for us, complete with reagents, but I haven't really heard from him soooo I decided to go ahead take care of it.

Requesting review by @Iamgoofball, of course.

To-Do

- [x] Get Pictures up for you guys
- [x] Make it compile.
- [x] Test to ensure the game mode starts, ends, and otherwise functions properly
- [x] Test for bugs and other oddities
 - [x] Fix bugs
 - [x] Fix more bugs
 - [x] Fix yet even more bugs
- [x] Fix a runtime related to blob broadcast
- [x] Fix a runtime related to empty overminds

Either case, this is a major update and overhaul to our Blob gamemode, which is over a year out of date, at this stage. 

Pics

blobs

Major Base Changes to the Mode

- Overall player requirement upped from 15 to 30
 - Blob just doesn't handle or play well with lowpop
 - Player to blob-core ratio also increased (less blobs)
- AI no longer gets quarantine instantly; only gets this law when it comes time to nuke the station.

 - The AI law activating at roundstart really hinders the crew in fighting the blob; it means that mining is effectively castrated from ever being able to have an impact when helping out; it also means that maneuvering around in space to fight the blob is technically something the AI can wreck your world for.
- Nuke threshold is now time based instead of progress based.
 - The blob has 50 minutes to win once its bursts; at that stage a new command report in generated and the nuclear code is given out, in addition to the AI getting a zeroth law to lockdown the station and maintain the lockdown at all costs--up to and including all of its other laws to do so.
- Command report and information send for blob is less obvious; command staff will be informed first, then the crew via a global announcement.
 - It'll still be more obvious on our server than usual, due to no command report being generated normally.
- Automatic Gamma Level removed
 - You'll have to manually request this now, from administration, ICly.

Blob Mechanics Changes

- Blobs are now reagent based
 - when you burst you're given a randomly generated blob reagent. Reagent blobs not only take on the color of the reagent they're utilizing, but when it attacks people this reagent will apply a unique effect (depending on reagent) to the person hit.  Any minions the blob controls (ie: blob mobs) will also take on the properties of this reagent.

New Blob Abilities

- Split consciousness
 - A one time ability which attempts to generate a new blob overmind--it's incredibly expensive and double the amount of destruction you'll have to cause to the station before you win.
- Storage Blob
 - New blob block--this increases the total amount of resources that you can store
- Re-roll Chemical
 - Re-rolls for a completely new reagent for you and your minions to utilize.
- Blob-speak
 - Makes all your controlled minions state a unified message at the same time.
- Spawn Blobbernaught
 - This consumes a factory note in the process, but generates a powerful hostile mob that will grind people to pulp.
- Relocate core
 - relocates your core to another node of your choosing

Removed Blob Abilities

- Create new core
 - Functionally, split consciousness takes care of this now and has better consequences for utilizing it.

Balance

- Blobs can expand into mobs again
 - This applies the reagent and does a little brute damage to them; the damage dealt is nothing remotely like it used to be back in the triple-click->dead days; the blobs will only deal a paltry 5 damage. It also obeys click cooldown.
- Shield blobs nerfed slightly
 - they're no longer brute resistant, but have 75 health as opposed to 60.

Admins
- Adds in proper check antags for the blob
 - You can now see who's infected with the blob and a list of the overminds
 - Can see progress/status of how much the blob has expanded

Fixes

- Fixes a number of odd behaviors with rally spores being ineffective, at best.
- Factories are tied to their overmind
 - You won't be in a click-fest contest fighting over what your blob-zombies do vs what their blob-zombies do.
- Game should properly end if you try to cheese victory, as a blob, by bursting in space.

Performance

- Blobs appear to more consistently garbage collect properly, technically meaning less odd behaviors and less processing time dedicated to del()'ing them

### CODE REVIEW

**Comment by Iamgoofball** [comment]
if I hear a single "muh reels" out of any of you i'm going to make chemistry run on ion bonds

File: code/game/gamemodes/blob/blob.dm [hunk 1]
```diff
@@ -56,12 +58,37 @@ var/list/blob_nodes = list()
...
+
+/datum/game_mode/blob/proc/blobize(var/mob/living/carbon/human/blob)
+	var/datum/mind/blobmind = blob.mind
+	if(!istype(blobmind))
+		return 0
+	infected_crew += blobmind
+	blobmind.special_role = "Blob"
+	log_game("[blob.key] (ckey) has been selected as a Blob")
+	greet_blob(blobmind)
+	blob << "<span class='userdanger'>You feel very tired and bloated!  You don't have long before you burst!</span>"
```
Comment (Iamgoofball) [inline code/game/gamemodes/blob/blob.dm]: this needs to be way longer, port over tg's values for blob bursting and timers

File: code/game/gamemodes/blob/powers.dm [hunk 2]
```diff
@@ -267,8 +301,107 @@
...
+		return
+	var/obj/effect/blob/node/N = locate(/obj/effect/blob) in blob_nodes
+	if(!N)
+		src << "<span class='warning'>A node is required to birth your offspring...</span>"
+		return
+
+	if(!can_buy(100))
+		return
+
+	verbs -= /mob/camera/blob/verb/split_consciousness //we've used our split_consciousness
```
Comment (Iamgoofball) [inline code/game/gamemodes/blob/powers.dm]: MAKE SURE TO REMOVE THIS FROM THE OFFSPRING TOO

I tried enabling this on tg without removing from the offspring, it was horrifying.

File: code/modules/reagents/newchem/Blob-Reagents.dm [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	id = "b_omnizine"
+	description = ""
+	color = "#C8A5DC"
+	message = "The blob squirts something at you"
+	message_living = ", and you feel great"
+
+/datum/reagent/blob/omnizine/reaction_mob(var/mob/living/M as mob, var/method=TOUCH, var/volume)
+	if(method == TOUCH)
+		var/ratio = volume/25
+		M.reagents.add_reagent("omnizine", 11*ratio)
```
Comment (Iamgoofball) [inline code/modules/reagents/newchem/Blob-Reagents.dm]: before you say "how is this good for the blob"

2 words: omnizine overdose

**Comment by Iamgoofball** [comment]
Serious talk:

You aren't going to need the auto-nuke blob stuff.

Axe the quarantine shit entirely. In practice on TG it just gets annoying.

**Comment by Iamgoofball** [comment]
Serious Talk 2:

Players are going to need to Git Gud and Git Robust to fight a blob. I don't want to hear any "it's 2 hard!" because here's how you're gonna kick this blob's fleshy teeth in:

# EMITTERS EMITTERS EMITTERS
## EMITTERS ARE YOUR BEST FRIEND.
 Get emitters up and running near the blob ASAP. Aim it at the core. I don't care if your speshul snoflaek character doesn't understand how it works, just "look at your PDA for instructions" and help set the damn emitter up on the blob. Go shut off the engine and steal the emitters from the engine. You won't need it because when the blob dies the round ends, and if it doesn't die, the round ends anyways.
# GET DAT FUKKEN AUTOLATHE
## WELDERS ARE YOUR SECOND BEST FRIEND
Have R&D construct a forward autolathe near the blob IMMEDIATELY. Make sure they upgrade the damn thing too, so it spits welders out faster than a greytiding chemist amped up on meth, morphine, and ephedrine.
If you aren't anyone with good weapons, print yourself a welder and start hammering at that blob.
I don't care if your blue haired floor length braid medbay stutterwhore nurse doesn't know how the welder works, "look it up on your PDA" and then print a welder and start hammering at the blob.
### DO NOT BRING A WELDING TANK NEAR THE BLOB. EVER.
It is too risky to bring it near the blob. If you run out of fuel, go print a new welder.
It is faster, cheaper, and safer to print a new welder.

I'll add more to this list soon(tm).

**Comment by Flip-Sandwich** [comment]
so far:

AI law removal: :+1:^:100:, fuck that noise.
Reagents seem fun enough, variety'll be probably good, and hm
Relocating core is... iffy, I don't think that GTFO is really a good solution for... what is it a solution to anyway?


Will do more when I sleep

**Comment by Iamgoofball** [comment]
@Flip-Sandwich It's pretty handy on TG, I use it all the time as a blob.

File: code/game/gamemodes/blob/powers.dm (hunk 2 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/blob/powers.dm]: Yup; the check is in place to prevent them from having it.

**Comment by Fox-McCloud** [comment]
@Flip-Sandwich 

Core re-locate is similar to AI shunting--it's a way to protect yourself by shifting around (for a cost). Technically AI shunting isn't possible at the moment because NanoUI doesn't support it--but that's a different subject.

**Comment by Earthdivine** [comment]
Core relocatin is as bull as AI Shaunting, unless it's only nodes and not any of the existing blob. Even then, that's something that will need to be carefully watched and maybe need axing since that makes Blob into hide and seek rather then actually fighting it.

Same goes with attackin mobs. It was never the damage it does, it's been the fact it's damage equal to your resources. However I don't know how the resource system works now so that's a moot point to argue on. 

Though, might I ask what's the point of keeping the fire resistance now that Plasma filling the rooms is not an easy thing to do anymore? It would give a use to the forgotten Flamethrowers and Incidary Grenades which have very little use now a days.

**Comment by Fox-McCloud** [comment]
@Earthdivine 

- You can only relocate to nodes.
- Blobs being able to attack mobs is pretty much *required* under this rework or the blob will just sit there doing nothing most of the time.
- It's not exactly rocket since, involving, or skillful to open up a can of plasma, light it on fire, and have it kill most of the blob because of it--it's meant to be a crew vs blob game, not one man opens a plasma canister and wins game. =p

File: code/game/gamemodes/blob/blob_finish.dm [hunk 4]
```diff
@@ -1,78 +1,42 @@
...
-		feedback_set_details("round_end_result","win - blob eliminated")
+		feedback_set_details("round_end_result","loss - blob eliminated")
 		world << "<FONT size = 3><B>The staff has won!</B></FONT>"
 		world << "<B>The alien organism has been eradicated from the station</B>"
-
-		var/datum/station_state/end_state = new /datum/station_state()
-		end_state.count()
-		var/percent = round( 100.0 *  start_state.score(end_state), 0.1)
-		world << "<B>The station is [percent]% intact.</B>"
-		log_game("Blob mode was won with station [percent]% intact.")
```
Comment (Robertha89) [inline code/game/gamemodes/blob/blob_finish.dm]: but seeing how much of the station the blob ruined was also part of the round, not sure why this is removed.

File: code/game/gamemodes/blob/blobs/core.dm [hunk 5]
```diff
@@ -1,93 +1,116 @@
...
-				B.verbs += /mob/camera/blob/proc/create_core
-			if(istype(ticker.mode, /datum/game_mode/blob))
-				var/datum/game_mode/blob/mode = ticker.mode
-				mode.infected_crew += B.mind
-			return 1
-		return 0
+	overmind_get_delay = world.time + 300 // 30 seconds
+
+	if(overmind)
+		qdel(overmind)
```
Comment (Robertha89) [inline code/game/gamemodes/blob/blobs/core.dm]: With this it's somewhat harder to insert a person into a blob core without a mind.

**Comment by Robertha89** [comment]
fire is still really effective, saddly fired is nerfed, also Durand with Ultra AC 2 defeats a blob every time without fault.
Also, all the blob reagents lack process_flags. Not sure if it changes anything, but ipc's being immune to this seems kinda silly.

**Comment by Iamgoofball** [comment]
@Robertha89 They don't apply on mob effects. They work via reaction_mob(), so IPCs still take the damage.

File: code/game/gamemodes/blob/blobs/core.dm (hunk 5 above)
Comment (Iamgoofball) [inline code/game/gamemodes/blob/blobs/core.dm]: Making a new blob core SHOULD be making a new overmind with it in the new.

File: code/game/gamemodes/blob/blob_finish.dm (hunk 4 above)
Comment (Iamgoofball) [inline code/game/gamemodes/blob/blob_finish.dm]: Yeah, this needs re-added.

**Comment by Earthdivine** [comment]
Except fire proof blob was one of THE most overpowered tools the blob has. There's a reason I often told new blobs to start in the Incinerator, with ZAS it was a way to pop off a fire to buy yourself time to mass up resources and make more nodes. Plus with LINDA popping a canister off won't be as destructive now, particularly since it would still harm the crew and melt the floor.  It seems like a small thin to worry about to not include one of the things that SHOULD work on the blob, and is one of the few Crowd Control weapons in the game, which helps so much in blob

**Comment by Iamgoofball** [comment]
@Earthdivine You have so, so, so many ways to deal with a blob. Flashbangs, Welders, Emitters, bombs, anything that does damage.

This PR is a 1:1 replica of TG blob.

TG players have the same weapons you guys do against a blob.

If they can handle blobs without fire, you can too.

**Comment by ZomgPonies** [comment]
I believe part of the issue is being able to 'whip up a hellmix in 5 seconds' in the first place. Power should come with time. Powerful things should require time and effort to gain. Not sekrit squirrel club recipes or knowledge.

**Comment by Earthdivine** [comment]
Don't balance or make arguments based on other stations please. We are not TG.

**Comment by Iamgoofball** [comment]
@Earthdivine 
You didn't read what I said.

**Comment by Earthdivine** [comment]
I only addressed part of it since ponies had addressed the other part. Infact half of that is say "TG  can do this". The blob is the most counter intuitive mob we have on game, and one of the weapons that SHOULD work is also one of the least used ever. If the issue is allowing Burn mixes to work that's more in issue with Chem then blob, as it was given that immunity to protect it from AIs instant killing it with Atmos

**Comment by Fox-McCloud** [comment]
Fixed one bug with overmind chat.

Still need to pin down two runtimes, then it's pretty much just testing and re-testing edge cases.

**Comment by Iamgoofball** [comment]
@Earthdivine 
You still didn't read what I said.

You're inheriting TG code, and TG tactics. You don't need fire against blobs. You have all the tools you need already. Use them.

File: code/game/gamemodes/blob/blobs/core.dm (hunk 5 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/blob/blobs/core.dm]: And indeed, it does create a new overmind in new()

File: code/game/gamemodes/blob/powers.dm [hunk 6]
```diff
@@ -267,8 +301,107 @@
...
+	set desc = "Speak with your blob spores and blobbernauts as your mouthpieces. This action is free."
+
+	var/speak_text = input(usr, "What would you like to say with your minions?", "Blob Broadcast", null) as text
+
+	if(!speak_text)
+		return
+	else
+		usr << "You broadcast with your minions, <B>[speak_text]</B>"
+	for(var/mob/living/simple_animal/hostile/blob_minion in blob_mobs)
+		blob_minion.say(speak_text)
```
Comment (Fox-McCloud) [inline code/game/gamemodes/blob/powers.dm]: whenever blob mobs say something and mob/camera/blob hears it, it runtimes--empty pick from list relevated to hear_say() currently what I'm trying to pin down.

File: code/game/gamemodes/blob/blob_finish.dm (hunk 4 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/blob/blob_finish.dm]: Yeah, can be done.

**Comment by IcyV** [comment]
Blob is already way too easy to win, I agree with any and all changes

**Comment by Fox-McCloud** [comment]
Wellp, that's interesting. Runtime on split consciousness is also present on /TG/ *wags a finger at!*

**Comment by Fox-McCloud** [comment]
I didn't feel like re-adding station integrity for blob.....

So I added it for every single game mode.

**Comment by Earthdivine** [comment]
I apologize, I saw Hellburn and thought chemical grenades rather then atmos mix.

**Comment by Robertha89** [comment]
@Fox-McCloud You took the best part of blob end, and made in universal... This is amazing.

**Comment by Aurorablade** [comment]
"so the round is over..and heres how much you messed up the station in the process!'

**Comment by KasparoVy** [comment]
"The AI law activating at roundstart really hinders the crew in fighting the blob; it means that mining is effectively castrated from ever being able to have an impact when helping out; it also means that maneuvering around in space to fight the blob is technically something the AI can wreck your world for."

As someone who almost exclusively mines, this busts my nuts :'(

@Iamgoofball Your Serious Talk 2 post is 10/10 and should be quoted and reposted on the para forums lmfao

File: code/_globalvars/misc.dm [hunk 7]
```diff
@@ -62,4 +62,6 @@ var/score_richestkey = null
 var/score_dmgestname = null // who had the most damage on the shuttle (but was still alive)
 var/score_dmgestjob = null
 var/score_dmgestdamage = 0
-var/score_dmgestkey = null
\ No newline at end of file
+var/score_dmgestkey = null
+
+var/TAB = "&nbsp;&nbsp;&nbsp;&nbsp;"
```
Comment (tigercat2000) [inline code/_globalvars/misc.dm]: ... will this even work? Is byond/internet explorer really shit enough to not suppress white-spaces?

File: code/game/gamemodes/blob/blob.dm [hunk 8]
```diff
@@ -136,30 +164,29 @@ var/list/blob_nodes = list()
...
 		stage(1)
 
 		// Stage 2
-		while(blobs.len < blobwincount*stage_2_threshold)
-			sleep(10)
+		sleep(30000)
 		stage(2)
 
-	..()
+	return ..(0)
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blob.dm]: Um. proc/post_setup doesn't take any arguments.

File: code/game/gamemodes/blob/blob_report.dm [hunk 9]
```diff
@@ -2,17 +2,16 @@
...
 			return
 		if(1)
-			interceptname = "Biohazard Alert"
-			intercepttext += "<FONT size = 3><B>Nanotrasen Update</B>: Biohazard Alert.</FONT><HR>"
+			intercepttext += "<FONT size = 3><B>NanoTrasen Update</B>: Biohazard Alert.</FONT><HR>"
 			intercepttext += "Reports indicate the probable transfer of a biohazardous agent onto [station_name()] during the last crew deployment cycle.<BR>"
-			intercepttext += "Preliminary analysis of the organism classifies it as a level 7 biohazard. Its origin is unknown.<BR>"
-			intercepttext += "Nanotrasen has issued a directive 7-10 for [station_name()]. The station is to be considered quarantined.<BR>"
+			intercepttext += "Preliminary analysis of the organism classifies it as a level 5 biohazard. Its origin is unknown.<BR>"
+			intercepttext += "NanoTrasen has issued a directive 7-10 for [station_name()]. The station is to be considered quarantined.<BR>"
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blob_report.dm]: The overwhelming majority of "Nanotrasen" in our code is just "Nanotrasen", not "NanoTrasen"

File: code/game/gamemodes/blob/blob_report.dm [hunk 10]
```diff
@@ -42,13 +46,12 @@
 		comm.messagetext.Add(intercepttext)
 		if(!(comm.stat & (BROKEN | NOPOWER)) && comm.prints_intercept)
 			var/obj/item/weapon/paper/intercept = new /obj/item/weapon/paper( comm.loc )
-			intercept.name = "paper- [interceptname]"
+			intercept.name = "Classified Central Command Update"
 			intercept.info = intercepttext
-	world << sound('sound/AI/commandreport.ogg')
+	command_announcement.Announce("A report has been downloaded and printed out at all communications consoles.", "Incoming Classified Message", 'sound/AI/commandreport.ogg');
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blob_report.dm]: What's with the random semicolon at the end of this?

File: code/game/gamemodes/blob/blobs/blob_mobs.dm [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+	melee_damage_upper = 15
+	icon = H.icon
+	speak_emote = list("groans")
+	icon_state = "zombie2_s"
+	H.h_style = null
+	H.update_hair()
+	human_overlays = H.overlays
+	update_icons()
+	H.loc = src
+	loc.visible_message("<span class='warning'>The corpse of [H.name] suddenly rises!</span>")
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: loc. is probably unneccesary here

File: code/game/gamemodes/blob/blobs/blob_mobs.dm [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+	ghostize()
+	qdel(src)
+
+/mob/living/simple_animal/hostile/blob/blobspore/Destroy()
+	if(factory)
+		factory.spores -= src
+	factory = null
+	if(contents)
+		for(var/mob/M in contents)
+			M.loc = src.loc
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blobs/blob_mobs.dm]: using get_turf(src) would probably be saner here

File: code/game/gamemodes/blob/blobs/core.dm [hunk 13]
```diff
@@ -1,93 +1,116 @@
...
+		var/mob/camera/blob/B = new(src.loc)
+		B.key = C.key
+		B.blob_core = src
+		src.overmind = B
+		color = overmind.blob_reagent_datum.color
+		if(B.mind && !B.mind.special_role)
+			B.mind.special_role = "Blob Overmind"
+		spawn(0)
+			if(is_offspring)
+				B.verbs -= /mob/camera/blob/verb/split_consciousness
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blobs/core.dm]: So, uh, a blob can only bring one other player in, and that's it? That's... I'm not sure that's a good thing.

File: code/game/gamemodes/blob/blobs/shield.dm [hunk 14]
```diff
@@ -3,18 +3,19 @@
...
-	update_icon()
-		if(health <= 0)
-			playsound(get_turf(src), 'sound/effects/splat.ogg', 50, 1)
-			Delete()
-			return
+/obj/effect/blob/shield/update_icon()
+	if(health <= 0)
+		qdel(src)
 		return
+	return
```
Comment (tigercat2000) [inline code/game/gamemodes/blob/blobs/shield.dm]: I'm 99% sure this is a pointless return.

File: code/game/gamemodes/blob/blob_report.dm (hunk 9 above)
Comment (TheDZD) [inline code/game/gamemodes/blob/blob_report.dm]: Yup, pretty sure someone went through and changed most instances of NanoTrasen to Nanotrasen a good while ago.

File: code/game/gamemodes/blob/blobs/core.dm (hunk 13 above)
Comment (Iamgoofball) [inline code/game/gamemodes/blob/blobs/core.dm]: Noope, trust me, it needs this.

Tg blobs used to have offspring splitting, it ended with RAPID MITOSIS BLOB.

Balance shows this is needed.

Comment (Robertha89) [inline code/game/gamemodes/blob/blobs/core.dm]: considering this is a valid strategy of blob in general, make a billion blob cores, I see no issue considering admins often give ERT access to fucking pulse weapons that kills cores in like 4 shots, not to forget about C4'ing cores.

Comment (Iamgoofball) [inline code/game/gamemodes/blob/blobs/core.dm]: No, that fucks over the balance of blob.

It isn't happening.

Comment (Fox-McCloud) [inline code/game/gamemodes/blob/blobs/core.dm]: As someone who briefly witnesses the aftermath of 'rapid mitosis"...yeah, it's definitely necessary.

File: code/_globalvars/misc.dm (hunk 7 above)
Comment (Fox-McCloud) [inline code/_globalvars/misc.dm]: Yes, yes it does.

**Comment by Fox-McCloud** [comment]
Fixed the runtime with blob broadcast--it also has the side effect of letting blob overminds understand what their own mobs are saying (this isn't a thing on TG).

Only one runtime left to go, then it's onto nosing out any other oddities...

**Comment by Fox-McCloud** [comment]
Well, as far as I can tell, it's really only the overmind runtime still left--everything I've tested so far seems to behave and function appropriately.

SOON.

**Comment by Fox-McCloud** [comment]
Ahha! Knew I was forgetting something; updated the blob event, as well.

**Comment by Fox-McCloud** [comment]
Fixed the runtime related to overmindless blob cores.

It means that blobs that spawn without overminds will end up being pure white, but when they do end up getting clients, that all resets.

This should now be ready to merge.

**Comment by TheDZD** [comment]
Wiki page is going to need updating after this.
